query/stdlib/influxdata/influxdb/rules.go

```diff
@@ -638,3 +639,81 @@ func (SortedPivotRule) Rewrite(ctx context.Context, pn plan.Node) (plan.Node, bo
 	}
 	return pn, false, nil
 }
+
+//
+// Push Down of window aggregates.
+// ReadRangePhys |> window |> { min, max, mean }
+//
+type PushDownWindowAggregateRule struct{}
+
+func (PushDownWindowAggregateRule) Name() string {
+	return "PushDownWindowAggregateRule"
+}
+
+func (rule PushDownWindowAggregateRule) Pattern() plan.Pattern {
+	return plan.OneOf(
+		[]plan.ProcedureKind{
+			universe.MinKind,
+			universe.MaxKind,
+			universe.MeanKind,
+		},
+		plan.Pat(universe.WindowKind, plan.Pat(ReadRangePhysKind)))
+}
+
+func (PushDownWindowAggregateRule) Rewrite(ctx context.Context, pn plan.Node) (plan.Node, bool, error) {
+	// FIXME: need to check capabilities.
+	// reader := GetStorageDependencies(ctx).FromDeps.Reader
+	// windowAggregateReade := reader.(reads.WindowAggregateReader)
+	// if !storeReader.HasWindowAggregateCapability(ctx) {
+	//  	return pn, false, nil
+	// }
+
+	// Check the aggregate function spec. Require operation on _value.
+	fnNode := pn
+	switch fnNode.Kind() {
+	case universe.MinKind:
+		minSpec := fnNode.ProcedureSpec().(*universe.MinProcedureSpec)
```

Comment:
It occurs to me that the specific functions that are supported in pushdowns will be specific to the storage nodes we're talking to.  Does the capabilities API expose which aggregate functions are supported?

In our initial update to storage, only one aggregate function will be supported.  It would be nice if, once we add support for other aggregate functions, those nodes begin reporting their new capabilities, and we don't have to make any changes to the planner rule to reflect that, it would just start working.

Reply:
Yeah if we assign the capabilities now we can definitely do that. @wolffcm